src/aws_encryption_sdk/keyrings/aws_kms/client_suppliers.py

```diff
@@ -0,0 +1,136 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""AWS KMS client suppliers for use with AWS KMS keyring.
+
+.. versionadded:: 1.5.0
+
+"""
+import logging
+
+import attr
+import six
+from attr.validators import deep_iterable, instance_of
+from botocore.client import BaseClient
+from botocore.session import Session as BotocoreSession
+
+from aws_encryption_sdk.exceptions import UnknownRegionError
+from aws_encryption_sdk.internal.validators import value_is_not_a_string
+
+from .client_cache import ClientCache
+
+try:  # Python 3.5.0 and 3.5.1 have incompatible typing modules
+    from typing import Any, Dict, Union  # noqa pylint: disable=unused-import
+except ImportError:  # pragma: no cover
+    # We only actually need these imports when running the mypy checks
+    pass
+
+_LOGGER = logging.getLogger(__name__)
+__all__ = ("ClientSupplier", "DefaultClientSupplier", "AllowRegionsClientSupplier", "DenyRegionsClientSupplier")
+
+
+class ClientSupplier(object):
+    """Base class for client suppliers.
+
+    .. versionadded:: 1.5.0
+
+    """
+
+    def client(self, region_name):
+        # type: (Union[None, str]) -> BaseClient
+        """Return a client for the requested region.
+
+        :rtype: BaseClient
+        """
+        raise NotImplementedError("'ClientSupplier' does not implement 'client'")
+
+
+@attr.s
+class DefaultClientSupplier(ClientSupplier):
```

Comment:
How do I build a supplier that has different options, like timeout?
I think that people generally will want to use the cache,
but is that always the case?

Reply:
addressed elsewhere.